osbs/build/user_params.py

```diff
@@ -113,55 +87,41 @@ def load_user_params_from_json(user_params_json):
-        self.component = BuildParam('component')
-        self.image_tag = BuildParam("image_tag")
-        self.koji_target = BuildParam("koji_target", allow_none=True)
-        self.koji_task_id = BuildParam('koji_task_id', allow_none=True)
-        self.platform = BuildParam("platform", allow_none=True)
-        self.orchestrator_deadline = BuildParam('orchestrator_deadline', allow_none=True)
-        self.reactor_config_map = BuildParam("reactor_config_map", allow_none=True)
-        self.reactor_config_override = BuildParam("reactor_config_override", allow_none=True)
-        self.scratch = BuildParam('scratch', allow_none=True)
-        self.signing_intent = BuildParam('signing_intent', allow_none=True)
-        self.user = UserParam()
-        self.worker_deadline = BuildParam('worker_deadline', allow_none=True)
-
-        self.required_params = [
-            self.build_json_dir,
-            self.koji_target,
-            self.user,
-        ]
-        self.convert_dict = {}
-
-    def attrs_finalizer(self):
-        for _, param in self.__dict__.items():
-            if isinstance(param, BuildParam):
-                # check that every parameter has a unique name
-                if param.name in self.convert_dict:
-                    raise OsbsValidationException(
-                        'Two user params with the same name: {}'.format(param.name))
-                self.convert_dict[param.name] = param
+class BuildCommon(BuildParamsBase):
+    """Common user parameters, class should be considered abstract"""
+
+    # Must be defined in subclasses
+    KIND = NotImplemented
+
+    arrangement_version = BuildParam("arrangement_version",
+                                     default=REACTOR_CONFIG_ARRANGEMENT_VERSION)
+    # build_from contains the full build_from string, including the source type prefix
+    build_from = BuildParam("build_from")
+    # build_image contains the buildroot name, whether the buildroot is a straight image or an
+    # imagestream.  buildroot_is_imagestream indicates what type of buildroot
+    build_image = BuildParam("build_image")
+    buildroot_is_imagestream = BuildParam("buildroot_is_imagestream", default=False)
+    build_json_dir = BuildParam("build_json_dir", required=True)
+    component = BuildParam("component")
+    image_tag = BuildParam("image_tag")
+    koji_target = BuildParam("koji_target")
+    koji_task_id = BuildParam("koji_task_id")
+    platform = BuildParam("platform")
+    orchestrator_deadline = BuildParam("orchestrator_deadline")
+    reactor_config_map = BuildParam("reactor_config_map")
+    reactor_config_override = BuildParam("reactor_config_override")
+    scratch = BuildParam("scratch")
+    signing_intent = BuildParam("signing_intent")
+    user = UserParam(required=True)
+    worker_deadline = BuildParam("worker_deadline")
+
+    def __init__(self, build_json_dir=None, **kwargs):
+        # TODO: why is is build_json_dir special?
```

Comment:
Was there any specific reason why `build_json_dir` had to be set in `__init__()` rather than together with the other params?

Reply:
After refactoring set_params, it does not seem to be the case